src/prommis/nanofiltration/costing/diafiltration.py

```diff
@@ -0,0 +1,713 @@
+    MSContactor,
+    MSContactorInitializer,
+    Mixer,
+    MixingType,
+    MomentumMixingType,
+    MixerInitializer,
+)
+
+
+class _StateBlock(StateBlock):
+    def fix_initialization_states(self):
+        """
+        Fixes state variables for state blocks.
+
+        Returns:
+            None
+        """
+        fix_state_vars(self)
+
+    def initialization_routine(self):
+        pass
+
+
+@declare_process_block_class("LiCoStateBlock", block_class=_StateBlock)
+class LiCoStateBlock1Data(StateBlockData):
+    def build(self):
+        super().build()
+
+        self.flow_vol = pyo.Var(
+            units=pyo.units.m**3 / pyo.units.hour,
+            bounds=(1e-8, None),
+        )
+        self.conc_mass_solute = pyo.Var(
+            ["Li", "Co"],
+            units=pyo.units.kg / pyo.units.m**3,
+            bounds=(1e-8, None),
+        )
+
+    def get_material_flow_terms(self, p, j):
+        if j == "solvent":
+            # Assume constant density of pure water
+            return self.flow_vol * self.params.dens_H2O
+        else:
+            return self.flow_vol * self.conc_mass_solute[j]
+
+    def get_material_flow_basis(self):
+        return MaterialFlowBasis.mass
+
+    def define_state_vars(self):
+        return {
+            "flow_vol": self.flow_vol,
+            "conc_mass_solute": self.conc_mass_solute,
+        }
+
+
+@declare_process_block_class("LiCoParameters")
+class LiCoParameterData(PhysicalParameterBlock):
+    def build(self):
+        super().build()
+
+        self.phase1 = Phase()
+
+        self.solvent = Component()
+        self.Li = Component()
+        self.Co = Component()
+
+        self.dens_H2O = pyo.Param(
+            default=1000,
+            units=pyo.units.kg / pyo.units.m**3,
+        )
+
+        self._state_block_class = LiCoStateBlock
+
+    @classmethod
+    def define_metadata(cls, obj):
+        obj.add_default_units(
+            {
+                "time": pyo.units.hour,
+                "length": pyo.units.m,
+                "mass": pyo.units.kg,
+                "amount": pyo.units.mol,
+                "temperature": pyo.units.K,
+            }
+        )
+
+
+# Global constants
+Jw = 0.1 * pyo.units.m / pyo.units.hour
+w = 1.5 * pyo.units.m
+diff_Co = 2.64e-06  # m2/h   https://www.aqion.de/site/diffusion-coefficients
+diafiltrate_inlet_pressure = 101325 * pyo.units.Pa
+diafiltrate_outlet_pressure = 145 * pyo.units.psi
+Q_feed = 100 * pyo.units.m**3 / pyo.units.h
+C_Li_feed = 1.7 * pyo.units.kg / pyo.units.m**3
+C_Co_feed = 17 * pyo.units.kg / pyo.units.m**3
+Q_diaf = 30 * pyo.units.m**3 / pyo.units.h
+C_Li_diaf = 0.1 * pyo.units.kg / pyo.units.m**3
+C_Co_diaf = 0.2 * pyo.units.kg / pyo.units.m**3
+
+
+def main():
+    m = build_and_init_model()
+
+    (
+        R_Li,
+        R_Co,
+        P_Li,
+        P_Co,
+        membrane_length,
+        feed_flow_rate,
+        diafiltrate_flow_rate,
+        permeate_flow_rate,
+    ) = get_values(m)
```

Comment:
An inline comment to explain what you are doing here would help. Also, a slightly more descriptive method name might also help clarify what is happening.